packages/menu/src/useMenu.ts

```diff
@@ -0,0 +1,644 @@
+                : {})
+            }
+          });
+        }
+
+        onChange({
+          type: changeType,
+          isExpanded: false,
+          ...(nextSelection ? { selectedItems: nextSelection } : {})
+        });
+      };
+
+      const handleItemKeyDown = (event: KeyboardEvent) => {
+        const { key } = event;
+        const isJumpKey = [KEYS.HOME, KEYS.END].includes(key);
+        const isSelectKey = [KEYS.SPACE, KEYS.ENTER].includes(key);
+        const isArrowKey = [KEYS.UP, KEYS.DOWN, KEYS.LEFT, KEYS.RIGHT].includes(key);
+        const isAlphanumericChar = key.length === 1 && /\S/u.test(key);
+
+        let changeType;
+        let payload;
+        let changes;
+
+        if (isSelectKey) {
+          event.preventDefault();
+
+          changeType = StateChangeTypes[toMenuItemKeyDownType(key)];
+          const nextSelection = getSelectionValue({ value, type, name, label, selected });
+
+          if (!isExpandedControlled || !isSelectionValueControlled) {
+            payload = {
+              ...(isExpandedControlled ? {} : { isExpanded: false }),
+              ...(!isSelectionValueControlled && nextSelection
+                ? { selectedItems: nextSelection }
+                : {})
+            };
+          }
+
+          changes = {
+            isExpanded: false,
+            ...(nextSelection ? { selectedItems: nextSelection } : {})
+          };
+        }
+
+        if (isArrowKey || isJumpKey || isAlphanumericChar) {
+          event.preventDefault();
+
+          changeType = isAlphanumericChar
+            ? StateChangeTypes.MenuItemKeyDown
+            : StateChangeTypes[toMenuItemKeyDownType(key)];
+          const nextFocusedValue = getNextFocusedValue(value, key, isAlphanumericChar);
+
+          payload = {
+            ...(isFocusedValueControlled ? {} : { focusedValue: nextFocusedValue })
+          };
+
+          changes = { focusedValue: nextFocusedValue };
+        }
+
+        if (changeType) {
+          payload && dispatch({ type: changeType, payload });
+
+          onChange({ type: changeType, ...changes });
+        }
+      };
+
+      const handleItemMouseEnter = () => {
+        const changeType = StateChangeTypes.MenuItemMouseMove;
+
+        dispatch({
+          type: changeType,
+          payload: {
+            ...(isFocusedValueControlled ? {} : { focusedValue: value })
+          }
+        });
+
+        onChange({
+          type: changeType,
+          focusedValue: value
+        });
+      };
+
+      const itemProps = getElementProps({
+        value: value as any,
+        ...elementProps,
+        onClick: composeEventHandlers(onClick, handleItemClick),
+        onKeyDown: composeEventHandlers(onKeyDown, handleItemKeyDown),
+        onMouseEnter: composeEventHandlers(onMouseEnter, handleItemMouseEnter)
+      });
+
+      if (itemProps.ref !== itemRefs[value]) {
+        itemRefs[value] = itemProps.ref as RefObject<any>;
```

Comment:
```suggestion
        itemRefs[value] = itemProps.ref as unknown as RefObject<T>;
```

slight type improvement over `any`?

Reply:
These are menu item refs - I don't think they have overlap with the Trigger type?